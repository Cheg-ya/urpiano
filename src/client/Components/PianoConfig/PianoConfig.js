import MusyngNames from 'soundfont-player/names/musyngkite.json';
import React, { Component } from 'react';
import { MidiNumbers } from 'react-piano';
import PropTypes from 'prop-types';
import './PianoConfig.scss';

class PianoConfig extends Component {
  constructor(props) {
    super(props);

    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.changeLastNote = this.changeLastNote.bind(this);
    this.changeFirstNote = this.changeFirstNote.bind(this);
    this.changeInstrument = this.changeInstrument.bind(this);
    this.createFirstNoteOptions = this.createFirstNoteOptions.bind(this);
    this.createInstrumentOptions = this.createInstrumentOptions.bind(this);
  }

  static defaultProps = {
    setConfig: () => {}
  };

  componentDidMount() {
    window.addEventListener('keydown', this.handleKeyDown);
  }

  componentWillUnmount() {
    window.removeEventListener('keydown', this.handleKeyDown);
  }

  handleKeyDown(e) {
    const { keyboardShortcutOffset, noteRange } = this.props;
    const totalKeyboardNum = noteRange.last - noteRange.first + 1;
    const minOffset = 0;
    const maxOffset = totalKeyboardNum - 18;

    if (e.key === 'ArrowLeft') {
      const reducedOffset = keyboardShortcutOffset - 1;
      
      if (reducedOffset >= minOffset) {
        this.props.setConfig({
          keyboardShortcutOffset: reducedOffset
        });
      }
    }

    if (e.key === 'ArrowRight') {
      const increasedOffset = keyboardShortcutOffset + 1;

      if (increasedOffset <= maxOffset) {
        this.props.setConfig({
          keyboardShortcutOffset: increasedOffset
        });
      }
    }
  }

  createInstrumentOptions() {
    return MusyngNames.map((instrumentName, i) => {
      return (<option key={i}>{instrumentName}</option>);
    });
  }

  createFirstNoteOptions() {
    const lastNote = this.props.noteRange.last;

    return MidiNumbers.NATURAL_MIDI_NUMBERS.map(number => {
      return (
        <option key={number} value={number} disabled={number >= (lastNote - 28)}>
          {MidiNumbers.getAttributes(number).note}
        </option>)
      });
  }

  createLastNoteOptions() {
    const firstNote = this.props.noteRange.first;

    return MidiNumbers.NATURAL_MIDI_NUMBERS.map(number => {
      return (
        <option key={number} value={number} disabled={number <= (firstNote + 28)}>
          {MidiNumbers.getAttributes(number).note}
        </option>)
      });
  }

  changeInstrument(e) {
    this.props.setConfig({
      instrumentName: e.target.value
    });
  }

  changeFirstNote(e) {
    const { noteRange } = this.props;
    const firstNote = parseInt(e.target.value);

    this.props.setConfig({
      noteRange: {
        first: firstNote,
        last: noteRange.last
      }
    });
  }

  changeLastNote(e) {
    const { noteRange } = this.props;
    const lastNote = parseInt(e.target.value);

    this.props.setConfig({
      noteRange: {
        first: noteRange.first,
        last: lastNote
      }
    });
  }

  render() {
    const { instrumentName, noteRange } = this.props;
    const firstNote = noteRange.first;
    const lastNote = noteRange.last;

    return (
      <div className="configContainer">
        <div className="configCover">
          <div className="firstNoteCover">
            <small>First Note</small>
            <select value={firstNote} onChange={this.changeFirstNote}>
              {this.createFirstNoteOptions()}
            </select>
          </div>
          <div className="lastNoteCover">
            <small>Last Note</small>
            <select value={lastNote} onChange={this.changeLastNote}>
              {this.createLastNoteOptions()}
            </select>
          </div>
          <div className="instrumentOptionCover">
            <small>Instrument</small>
            <select onChange={this.changeInstrument} value={instrumentName}>
              {this.createInstrumentOptions()}
            </select>
          </div>
        </div>
      </div>
    );
  }
}

export default PianoConfig;

PianoConfig.propTypes = {
  instrumentName: PropTypes.string.isRequired,
  keyboardShortcutOffset: PropTypes.number,
  noteRange: PropTypes.object.isRequired,
  setConfig: PropTypes.func.isRequired
};
