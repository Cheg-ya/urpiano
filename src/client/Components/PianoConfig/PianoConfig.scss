.configContainer {
  margin-top: 10px;
  padding: 10px;

  .configCover {
    div {
      display: inline-block;
    }

    small {
      font-size: 0.8rem;
      color: white;
      display: block;
    }

    select {
      font-size: 1vw;
      padding: 6px 12px;
      margin-top: 5px;
      border-radius: 4px;
      outline: none;
      transition: 0.5s;
    }

    select:focus {
      box-shadow: 0px 0px 1px 3px #a9daee;
    }

    .firstNoteCover, .lastNoteCover, .instrumentOptionCover {
      margin: 0 20px;
    }
  }
}
