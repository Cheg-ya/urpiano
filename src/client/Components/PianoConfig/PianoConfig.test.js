import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import PianoConfig from './PianoConfig';

describe('PianoConfig component test', () => {
  const instrument = 'acoustic_grand_piano';
  const keyboardShortcutOffset = 0;
  const noteRange = {
    first: 1,
    last: 30
  };

  const setConfig = jest.fn().mockImplementation(() => true);

  describe('render and props test', () => {
    const component = shallow(
      <PianoConfig 
        instrumentName={instrument}
        noteRange={noteRange}
        keyboardShortcutOffset={keyboardShortcutOffset}
        setConfig={setConfig}
      />
    );
  
    const instance = component.instance();
  
    it('renders correctly and check if state exists', () => {
      const tree = renderer.create(
        <PianoConfig 
          instrumentName={instrument}
          noteRange={noteRange}
          keyboardShortcutOffset={keyboardShortcutOffset}
          setConfig={setConfig}
        />
      ).toJSON();
  
      expect(tree).toMatchSnapshot();
      expect(instance).toBeInstanceOf(PianoConfig);
      expect(component.exists()).toBe(true);
      expect(component.exists('.configContainer')).toBe(true);
      expect(component.type()).toEqual('div');
      expect(component.state()).toBeNull();
    });
  
    it('checks props object\'s properties', () => {
      expect(instance.props).toHaveProperty('noteRange');
      expect(instance.props).toHaveProperty('setConfig');
      expect(instance.props).toHaveProperty('instrumentName');
      expect(instance.props).toHaveProperty('keyboardShortcutOffset');
      expect(instance.props.setConfig).toBe(setConfig);
      expect(instance.props.noteRange).toBe(noteRange);
      expect(instance.props.instrumentName).toBe(instrument);
      expect(instance.props.keyboardShortcutOffset).toBe(keyboardShortcutOffset);
    });

    it('checks props type', () => {
      expect(typeof instance.props.noteRange).toBe('object');
      expect(typeof instance.props.setConfig).toBe('function');
      expect(typeof instance.props.instrumentName).toBe('string');
      expect(typeof instance.props.keyboardShortcutOffset).toBe('number');
    });
  });

  describe('life-cycle method test', () => {
    it('calls componentDidMount & componentWillUnmount correctly', () => {
      const didMount = jest.spyOn(PianoConfig.prototype, 'componentDidMount');
      const eventCallback = jest.spyOn(PianoConfig.prototype, 'handleKeyDown');
      const addEventListener = window.addEventListener;
      const removeEventListener = window.removeEventListener;

      window.addEventListener = jest.fn().mockImplementation((e, cb) => cb(e));
      window.removeEventListener = jest.fn().mockImplementation((e, cb) => cb(e));

      const component = mount(<PianoConfig
        instrumentName={instrument}
        noteRange={noteRange}
        keyboardShortcutOffset={keyboardShortcutOffset}
        setConfig={setConfig}
      />);

      expect(didMount).toHaveBeenCalled();
      expect(didMount).toHaveBeenCalledTimes(1);
      expect(window.addEventListener).toHaveBeenCalled();
      expect(eventCallback).toHaveBeenCalled();
      expect(eventCallback).toHaveBeenCalledTimes(1);

      const unmount = jest.spyOn(PianoConfig.prototype, 'componentWillUnmount');
      component.unmount();

      expect(unmount).toHaveBeenCalled();
      expect(unmount).toHaveBeenCalledTimes(1);
      expect(window.removeEventListener).toHaveBeenCalled();
      expect(eventCallback).toHaveBeenCalled();
      expect(eventCallback).toHaveBeenCalledTimes(2);

      window.addEventListener = addEventListener;
      window.removeEventListener = removeEventListener;
    });
  });

  describe('component custom methods test', () => {
    it('tests methods that call right after render', () => {
      const createFirstNoteOptions = jest.spyOn(PianoConfig.prototype, 'createFirstNoteOptions');
      const createLastNoteOptions = jest.spyOn(PianoConfig.prototype, 'createLastNoteOptions');
      const createInstrumentOptions = jest.spyOn(PianoConfig.prototype, 'createInstrumentOptions');

      const component = mount(<PianoConfig
        instrumentName={instrument}
        noteRange={noteRange}
        keyboardShortcutOffset={keyboardShortcutOffset}
        setConfig={setConfig}
      />);
  
      expect(createFirstNoteOptions).toHaveBeenCalled();
      expect(createFirstNoteOptions).toHaveReturned();
      expect(createLastNoteOptions).toHaveBeenCalled();
      expect(createLastNoteOptions).toHaveReturned();
      expect(createInstrumentOptions).toHaveBeenCalled();
      expect(createInstrumentOptions).toHaveReturned();

      component.unmount();
    });

    it('calls props function when relative functions called', () => {
      const component = mount(<PianoConfig
        instrumentName={instrument}
        noteRange={noteRange}
        keyboardShortcutOffset={keyboardShortcutOffset}
        setConfig={setConfig}
      />);
  
      const props = component.props();
      const spy = jest.spyOn(props, 'setConfig');
  
      component.instance().changeFirstNote({ target: { value: 10 } });
      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveReturned();
  
      component.instance().changeLastNote({ target: { value: 10 } });
      expect(spy).toHaveBeenCalledTimes(2);
  
      component.instance().changeInstrument({ target: { value: 10 } });
      expect(spy).toHaveBeenCalledTimes(3);

      component.unmount();
    });

    it('calls methods when keydown event occurred', () => {
      const addEventListener = window.addEventListener;
      const removeEventListener = window.removeEventListener;

      const eventAndCallback = {};

      window.addEventListener = jest.fn(((e, cb) => eventAndCallback[e] = cb));
      window.removeEventListener = jest.fn(((e, cb) => eventAndCallback[e] = cb));

      const setConfig = jest.fn().mockImplementation(() => true);
      const component = mount(<PianoConfig
        instrumentName={instrument}
        noteRange={noteRange}
        keyboardShortcutOffset={keyboardShortcutOffset}
        setConfig={setConfig}
      />);

      eventAndCallback.keydown({ key: 'ArrowRight' });
      expect(setConfig).toHaveBeenCalled();
      eventAndCallback.keydown({ key: 'ArrowLeft' });
      expect(setConfig).toHaveBeenCalled();

      component.unmount();

      window.addEventListener = addEventListener;
      window.removeEventListener = removeEventListener;
    });
  });
});

