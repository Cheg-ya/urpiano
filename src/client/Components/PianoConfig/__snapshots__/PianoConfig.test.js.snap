// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PianoConfig component test render and props test renders correctly and check if state exists 1`] = `
<div
  className="configContainer"
>
  <div
    className="configCover"
  >
    <div
      className="firstNoteCover"
    >
      <small>
        First Note
      </small>
      <select
        onChange={[Function]}
        value={1}
      >
        <option
          disabled={true}
          value={12}
        >
          C0
        </option>
        <option
          disabled={true}
          value={14}
        >
          D0
        </option>
        <option
          disabled={true}
          value={16}
        >
          E0
        </option>
        <option
          disabled={true}
          value={17}
        >
          F0
        </option>
        <option
          disabled={true}
          value={19}
        >
          G0
        </option>
        <option
          disabled={true}
          value={21}
        >
          A0
        </option>
        <option
          disabled={true}
          value={23}
        >
          B0
        </option>
        <option
          disabled={true}
          value={24}
        >
          C1
        </option>
        <option
          disabled={true}
          value={26}
        >
          D1
        </option>
        <option
          disabled={true}
          value={28}
        >
          E1
        </option>
        <option
          disabled={true}
          value={29}
        >
          F1
        </option>
        <option
          disabled={true}
          value={31}
        >
          G1
        </option>
        <option
          disabled={true}
          value={33}
        >
          A1
        </option>
        <option
          disabled={true}
          value={35}
        >
          B1
        </option>
        <option
          disabled={true}
          value={36}
        >
          C2
        </option>
        <option
          disabled={true}
          value={38}
        >
          D2
        </option>
        <option
          disabled={true}
          value={40}
        >
          E2
        </option>
        <option
          disabled={true}
          value={41}
        >
          F2
        </option>
        <option
          disabled={true}
          value={43}
        >
          G2
        </option>
        <option
          disabled={true}
          value={45}
        >
          A2
        </option>
        <option
          disabled={true}
          value={47}
        >
          B2
        </option>
        <option
          disabled={true}
          value={48}
        >
          C3
        </option>
        <option
          disabled={true}
          value={50}
        >
          D3
        </option>
        <option
          disabled={true}
          value={52}
        >
          E3
        </option>
        <option
          disabled={true}
          value={53}
        >
          F3
        </option>
        <option
          disabled={true}
          value={55}
        >
          G3
        </option>
        <option
          disabled={true}
          value={57}
        >
          A3
        </option>
        <option
          disabled={true}
          value={59}
        >
          B3
        </option>
        <option
          disabled={true}
          value={60}
        >
          C4
        </option>
        <option
          disabled={true}
          value={62}
        >
          D4
        </option>
        <option
          disabled={true}
          value={64}
        >
          E4
        </option>
        <option
          disabled={true}
          value={65}
        >
          F4
        </option>
        <option
          disabled={true}
          value={67}
        >
          G4
        </option>
        <option
          disabled={true}
          value={69}
        >
          A4
        </option>
        <option
          disabled={true}
          value={71}
        >
          B4
        </option>
        <option
          disabled={true}
          value={72}
        >
          C5
        </option>
        <option
          disabled={true}
          value={74}
        >
          D5
        </option>
        <option
          disabled={true}
          value={76}
        >
          E5
        </option>
        <option
          disabled={true}
          value={77}
        >
          F5
        </option>
        <option
          disabled={true}
          value={79}
        >
          G5
        </option>
        <option
          disabled={true}
          value={81}
        >
          A5
        </option>
        <option
          disabled={true}
          value={83}
        >
          B5
        </option>
        <option
          disabled={true}
          value={84}
        >
          C6
        </option>
        <option
          disabled={true}
          value={86}
        >
          D6
        </option>
        <option
          disabled={true}
          value={88}
        >
          E6
        </option>
        <option
          disabled={true}
          value={89}
        >
          F6
        </option>
        <option
          disabled={true}
          value={91}
        >
          G6
        </option>
        <option
          disabled={true}
          value={93}
        >
          A6
        </option>
        <option
          disabled={true}
          value={95}
        >
          B6
        </option>
        <option
          disabled={true}
          value={96}
        >
          C7
        </option>
        <option
          disabled={true}
          value={98}
        >
          D7
        </option>
        <option
          disabled={true}
          value={100}
        >
          E7
        </option>
        <option
          disabled={true}
          value={101}
        >
          F7
        </option>
        <option
          disabled={true}
          value={103}
        >
          G7
        </option>
        <option
          disabled={true}
          value={105}
        >
          A7
        </option>
        <option
          disabled={true}
          value={107}
        >
          B7
        </option>
        <option
          disabled={true}
          value={108}
        >
          C8
        </option>
        <option
          disabled={true}
          value={110}
        >
          D8
        </option>
        <option
          disabled={true}
          value={112}
        >
          E8
        </option>
        <option
          disabled={true}
          value={113}
        >
          F8
        </option>
        <option
          disabled={true}
          value={115}
        >
          G8
        </option>
        <option
          disabled={true}
          value={117}
        >
          A8
        </option>
        <option
          disabled={true}
          value={119}
        >
          B8
        </option>
        <option
          disabled={true}
          value={120}
        >
          C9
        </option>
        <option
          disabled={true}
          value={122}
        >
          D9
        </option>
        <option
          disabled={true}
          value={124}
        >
          E9
        </option>
        <option
          disabled={true}
          value={125}
        >
          F9
        </option>
        <option
          disabled={true}
          value={127}
        >
          G9
        </option>
      </select>
    </div>
    <div
      className="lastNoteCover"
    >
      <small>
        Last Note
      </small>
      <select
        onChange={[Function]}
        value={30}
      >
        <option
          disabled={true}
          value={12}
        >
          C0
        </option>
        <option
          disabled={true}
          value={14}
        >
          D0
        </option>
        <option
          disabled={true}
          value={16}
        >
          E0
        </option>
        <option
          disabled={true}
          value={17}
        >
          F0
        </option>
        <option
          disabled={true}
          value={19}
        >
          G0
        </option>
        <option
          disabled={true}
          value={21}
        >
          A0
        </option>
        <option
          disabled={true}
          value={23}
        >
          B0
        </option>
        <option
          disabled={true}
          value={24}
        >
          C1
        </option>
        <option
          disabled={true}
          value={26}
        >
          D1
        </option>
        <option
          disabled={true}
          value={28}
        >
          E1
        </option>
        <option
          disabled={true}
          value={29}
        >
          F1
        </option>
        <option
          disabled={false}
          value={31}
        >
          G1
        </option>
        <option
          disabled={false}
          value={33}
        >
          A1
        </option>
        <option
          disabled={false}
          value={35}
        >
          B1
        </option>
        <option
          disabled={false}
          value={36}
        >
          C2
        </option>
        <option
          disabled={false}
          value={38}
        >
          D2
        </option>
        <option
          disabled={false}
          value={40}
        >
          E2
        </option>
        <option
          disabled={false}
          value={41}
        >
          F2
        </option>
        <option
          disabled={false}
          value={43}
        >
          G2
        </option>
        <option
          disabled={false}
          value={45}
        >
          A2
        </option>
        <option
          disabled={false}
          value={47}
        >
          B2
        </option>
        <option
          disabled={false}
          value={48}
        >
          C3
        </option>
        <option
          disabled={false}
          value={50}
        >
          D3
        </option>
        <option
          disabled={false}
          value={52}
        >
          E3
        </option>
        <option
          disabled={false}
          value={53}
        >
          F3
        </option>
        <option
          disabled={false}
          value={55}
        >
          G3
        </option>
        <option
          disabled={false}
          value={57}
        >
          A3
        </option>
        <option
          disabled={false}
          value={59}
        >
          B3
        </option>
        <option
          disabled={false}
          value={60}
        >
          C4
        </option>
        <option
          disabled={false}
          value={62}
        >
          D4
        </option>
        <option
          disabled={false}
          value={64}
        >
          E4
        </option>
        <option
          disabled={false}
          value={65}
        >
          F4
        </option>
        <option
          disabled={false}
          value={67}
        >
          G4
        </option>
        <option
          disabled={false}
          value={69}
        >
          A4
        </option>
        <option
          disabled={false}
          value={71}
        >
          B4
        </option>
        <option
          disabled={false}
          value={72}
        >
          C5
        </option>
        <option
          disabled={false}
          value={74}
        >
          D5
        </option>
        <option
          disabled={false}
          value={76}
        >
          E5
        </option>
        <option
          disabled={false}
          value={77}
        >
          F5
        </option>
        <option
          disabled={false}
          value={79}
        >
          G5
        </option>
        <option
          disabled={false}
          value={81}
        >
          A5
        </option>
        <option
          disabled={false}
          value={83}
        >
          B5
        </option>
        <option
          disabled={false}
          value={84}
        >
          C6
        </option>
        <option
          disabled={false}
          value={86}
        >
          D6
        </option>
        <option
          disabled={false}
          value={88}
        >
          E6
        </option>
        <option
          disabled={false}
          value={89}
        >
          F6
        </option>
        <option
          disabled={false}
          value={91}
        >
          G6
        </option>
        <option
          disabled={false}
          value={93}
        >
          A6
        </option>
        <option
          disabled={false}
          value={95}
        >
          B6
        </option>
        <option
          disabled={false}
          value={96}
        >
          C7
        </option>
        <option
          disabled={false}
          value={98}
        >
          D7
        </option>
        <option
          disabled={false}
          value={100}
        >
          E7
        </option>
        <option
          disabled={false}
          value={101}
        >
          F7
        </option>
        <option
          disabled={false}
          value={103}
        >
          G7
        </option>
        <option
          disabled={false}
          value={105}
        >
          A7
        </option>
        <option
          disabled={false}
          value={107}
        >
          B7
        </option>
        <option
          disabled={false}
          value={108}
        >
          C8
        </option>
        <option
          disabled={false}
          value={110}
        >
          D8
        </option>
        <option
          disabled={false}
          value={112}
        >
          E8
        </option>
        <option
          disabled={false}
          value={113}
        >
          F8
        </option>
        <option
          disabled={false}
          value={115}
        >
          G8
        </option>
        <option
          disabled={false}
          value={117}
        >
          A8
        </option>
        <option
          disabled={false}
          value={119}
        >
          B8
        </option>
        <option
          disabled={false}
          value={120}
        >
          C9
        </option>
        <option
          disabled={false}
          value={122}
        >
          D9
        </option>
        <option
          disabled={false}
          value={124}
        >
          E9
        </option>
        <option
          disabled={false}
          value={125}
        >
          F9
        </option>
        <option
          disabled={false}
          value={127}
        >
          G9
        </option>
      </select>
    </div>
    <div
      className="instrumentOptionCover"
    >
      <small>
        Instrument
      </small>
      <select
        onChange={[Function]}
        value="acoustic_grand_piano"
      >
        <option>
          accordion
        </option>
        <option>
          acoustic_bass
        </option>
        <option>
          acoustic_grand_piano
        </option>
        <option>
          acoustic_guitar_nylon
        </option>
        <option>
          acoustic_guitar_steel
        </option>
        <option>
          agogo
        </option>
        <option>
          alto_sax
        </option>
        <option>
          applause
        </option>
        <option>
          bagpipe
        </option>
        <option>
          banjo
        </option>
        <option>
          baritone_sax
        </option>
        <option>
          bassoon
        </option>
        <option>
          bird_tweet
        </option>
        <option>
          blown_bottle
        </option>
        <option>
          brass_section
        </option>
        <option>
          breath_noise
        </option>
        <option>
          bright_acoustic_piano
        </option>
        <option>
          celesta
        </option>
        <option>
          cello
        </option>
        <option>
          choir_aahs
        </option>
        <option>
          church_organ
        </option>
        <option>
          clarinet
        </option>
        <option>
          clavinet
        </option>
        <option>
          contrabass
        </option>
        <option>
          distortion_guitar
        </option>
        <option>
          drawbar_organ
        </option>
        <option>
          dulcimer
        </option>
        <option>
          electric_bass_finger
        </option>
        <option>
          electric_bass_pick
        </option>
        <option>
          electric_grand_piano
        </option>
        <option>
          electric_guitar_clean
        </option>
        <option>
          electric_guitar_jazz
        </option>
        <option>
          electric_guitar_muted
        </option>
        <option>
          electric_piano_1
        </option>
        <option>
          electric_piano_2
        </option>
        <option>
          english_horn
        </option>
        <option>
          fiddle
        </option>
        <option>
          flute
        </option>
        <option>
          french_horn
        </option>
        <option>
          fretless_bass
        </option>
        <option>
          fx_1_rain
        </option>
        <option>
          fx_2_soundtrack
        </option>
        <option>
          fx_3_crystal
        </option>
        <option>
          fx_4_atmosphere
        </option>
        <option>
          fx_5_brightness
        </option>
        <option>
          fx_6_goblins
        </option>
        <option>
          fx_7_echoes
        </option>
        <option>
          fx_8_scifi
        </option>
        <option>
          glockenspiel
        </option>
        <option>
          guitar_fret_noise
        </option>
        <option>
          guitar_harmonics
        </option>
        <option>
          gunshot
        </option>
        <option>
          harmonica
        </option>
        <option>
          harpsichord
        </option>
        <option>
          helicopter
        </option>
        <option>
          honkytonk_piano
        </option>
        <option>
          kalimba
        </option>
        <option>
          koto
        </option>
        <option>
          lead_1_square
        </option>
        <option>
          lead_2_sawtooth
        </option>
        <option>
          lead_3_calliope
        </option>
        <option>
          lead_4_chiff
        </option>
        <option>
          lead_5_charang
        </option>
        <option>
          lead_6_voice
        </option>
        <option>
          lead_7_fifths
        </option>
        <option>
          lead_8_bass__lead
        </option>
        <option>
          marimba
        </option>
        <option>
          melodic_tom
        </option>
        <option>
          music_box
        </option>
        <option>
          muted_trumpet
        </option>
        <option>
          oboe
        </option>
        <option>
          ocarina
        </option>
        <option>
          orchestra_hit
        </option>
        <option>
          orchestral_harp
        </option>
        <option>
          overdriven_guitar
        </option>
        <option>
          pad_1_new_age
        </option>
        <option>
          pad_2_warm
        </option>
        <option>
          pad_3_polysynth
        </option>
        <option>
          pad_4_choir
        </option>
        <option>
          pad_5_bowed
        </option>
        <option>
          pad_6_metallic
        </option>
        <option>
          pad_7_halo
        </option>
        <option>
          pad_8_sweep
        </option>
        <option>
          pan_flute
        </option>
        <option>
          percussive_organ
        </option>
        <option>
          piccolo
        </option>
        <option>
          pizzicato_strings
        </option>
        <option>
          recorder
        </option>
        <option>
          reed_organ
        </option>
        <option>
          reverse_cymbal
        </option>
        <option>
          rock_organ
        </option>
        <option>
          seashore
        </option>
        <option>
          shakuhachi
        </option>
        <option>
          shamisen
        </option>
        <option>
          shanai
        </option>
        <option>
          sitar
        </option>
        <option>
          slap_bass_1
        </option>
        <option>
          slap_bass_2
        </option>
        <option>
          soprano_sax
        </option>
        <option>
          steel_drums
        </option>
        <option>
          string_ensemble_1
        </option>
        <option>
          string_ensemble_2
        </option>
        <option>
          synth_bass_1
        </option>
        <option>
          synth_bass_2
        </option>
        <option>
          synth_brass_1
        </option>
        <option>
          synth_brass_2
        </option>
        <option>
          synth_choir
        </option>
        <option>
          synth_drum
        </option>
        <option>
          synth_strings_1
        </option>
        <option>
          synth_strings_2
        </option>
        <option>
          taiko_drum
        </option>
        <option>
          tango_accordion
        </option>
        <option>
          telephone_ring
        </option>
        <option>
          tenor_sax
        </option>
        <option>
          timpani
        </option>
        <option>
          tinkle_bell
        </option>
        <option>
          tremolo_strings
        </option>
        <option>
          trombone
        </option>
        <option>
          trumpet
        </option>
        <option>
          tuba
        </option>
        <option>
          tubular_bells
        </option>
        <option>
          vibraphone
        </option>
        <option>
          viola
        </option>
        <option>
          violin
        </option>
        <option>
          voice_oohs
        </option>
        <option>
          whistle
        </option>
        <option>
          woodblock
        </option>
        <option>
          xylophone
        </option>
      </select>
    </div>
  </div>
</div>
`;
